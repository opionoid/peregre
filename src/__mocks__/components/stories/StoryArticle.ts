import { IStoryArticleProps } from 'src/components/stories/StoryArticle';

const MOCK_STORY_ARTICLE_PROPS: IStoryArticleProps = {
  children: 'サンプル',
  title: 'タイトル',
  thumbnail: {
    src: 'https://picsum.photos/id/1037/400/300',
    alt: ''
  },
  publishDate: '2021年11月20日',
  author: 'ことれの',
  tags: ['和風'],
  difficulty: '標準',
  time: 4,
  people: {
    from: 2,
    to: 3
  },
  abstract: 'テキストが入ります。テキストが入ります。テキストが入ります。テキストが入ります。テキストが入ります。テキストが入ります。テキストが入ります。'
};

export default MOCK_STORY_ARTICLE_PROPS;
