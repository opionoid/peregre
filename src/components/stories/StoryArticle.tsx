
import React from 'react'
import { useToggle } from 'react-use'
import { Icons } from 'src/assets/icons'
import { space } from 'src/assets/style'
import { IDifficulty, IImage } from 'src/interfaces'
import styled from 'styled-components'
import { ButtonBase } from '../actor/button/ButtonBase'

export interface IStoryArticleProps {
  children: React.ReactNode
  title: string
  thumbnail: IImage
  publishDate: string // TODO: date-fns
  updateDate?: string
  author?: string
  tags: string[]
  difficulty: IDifficulty
  time: number
  people: {
    from: number
    to: number
  }
  abstract: string
}

export const StoryArticle: React.FC<IStoryArticleProps> = (props) => {
  const [shouldShowBody, toggleShown] = useToggle(false)
  const handleClick = () => toggleShown()
  return (
    <StoryArticleWrapper>
      <Thumbnail src={props.thumbnail.src} alt={props.thumbnail.alt || ''} />
      <Title>{props.title}</Title>
      <Author>{props.author || 'ことれの'}</Author>
      <InfoList>
        <Info><InfoIcon src={Icons.Person} decoding="async" />{props.people.from}~{props.people.to}</Info>
        <Info><InfoIcon src={Icons.Time} decoding="async" />{props.time}h</Info>
        <Info><InfoIcon src={Icons.MaxDepth} alt="難易度" decoding="async" />{props.difficulty}</Info>
      </InfoList>
      <Abstract>{props.abstract}</Abstract>
      <hr />
      <div aria-hidden={shouldShowBody}>
        <ButtonBase aria-controls="story-contents" onClick={handleClick}>書を読み進める</ButtonBase>
      </div>
      {/** TODO: 本当は文字が不安定になっているエフェクト入れたいけど P2 に回す */}
      <div id="story-contents" aria-hidden={!shouldShowBody}>{props.children}</div>
    </StoryArticleWrapper>
  )
}

const StoryArticleWrapper = styled.article`
margin-bottom: 9em;
  
&::before {
  content: '';
  position: fixed;
  z-index: -1;
  top: 0;
  left: 0;
  background-image: url('https://picsum.photos/1200/1200');
  background-attachment: fixed;
  background-repeat: no-repeat;
  background-position: 50% 50%;
  background-size: cover;
  width: 100%;
  height: 100vh;
  filter: opacity(0.07) grayscale(0.1) blur(8px) sepia(1);
}
`
const Thumbnail = styled.img`
margin-top: ${space.m};
width: 100%;
`
const Title = styled.h1`
margin-top: ${space.m};
`
const Author = styled.p`
text-align: end;
margin-right: 0.5rem;
`
const InfoList = styled.ul`
display: flex;
margin: 0 0 0 auto;
width: fit-content;
flex-wrap: wrap;
list-style: none;
`
const Info = styled.li`
display: flex;
align-items: center;
&:last-child {
  margin-right: 0.5rem;
}
&:not(:last-child) {
  margin-right: ${space.s};
}
> *:first-child {
  margin-right: ${space.xxs};
}
`
const InfoIcon = styled.img`
width: 1.625rem;
height: 1.625rem;
`
const Abstract = styled.p`
margin-top: ${space.m};
`
