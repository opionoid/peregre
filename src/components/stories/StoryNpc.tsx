
import React from 'react'
import { useToggle } from 'react-use'
import { Icons } from 'src/assets/icons'
import { space } from 'src/assets/style'
import { ISkill } from 'src/interfaces'
import { useBattle } from 'src/utils/hooks/useBattle'
import styled from 'styled-components'
import { ButtonBase } from '../actor/button/ButtonBase'

export interface IStoryNpcProps {
  name: string
  hp: number
  skills: Omit<ISkill, 'icon'>[]
}

export const StoryNpc: React.VFC<IStoryNpcProps> = (props) => {
  const [shouldShowSkills, toggleShown] = useToggle(false)

  const { onClickSkill } = useBattle()
  const [depth, setDepth] = React.useState<0 | 1 | 2 | 3 | 4>(2)
  const handleClick = (skill: ISkill) => onClickSkill(skill, setDepth, props.name)

  return (
    <StoryNpcWrapper>
      <Info>
        <h4>{props.name}</h4>
        <Status>
          <label aria-label="魂魄量"><Icon src={Icons.MaxHp} /></label>
          <HpInput />
        </Status>
        <Status>
          <label aria-label="魂魄量"><Icon src={Icons.MaxDepth} /></label>
          <span>{depth}</span>
        </Status>
        <ToggleButton>
          <ButtonBase aria-controls="skills" onClick={() => toggleShown()}>{shouldShowSkills ? 'スキル非表示' : 'スキル表示'}</ButtonBase>
        </ToggleButton>
      </Info>
      <div id="skills" aria-hidden={!shouldShowSkills}>
        {props.skills.map(skill => (
          <Skill>
            <SkillRow>
              <SkillName>{skill.name}</SkillName>
              <SkillDepth><Icon src={Icons.MaxDepth} />{skill.depth}</SkillDepth>
            </SkillRow>
            <SkillDescription>{skill.description}</SkillDescription>
            <SkillButton>
              <ButtonBase onClick={() => handleClick({ ...skill, icon: { src: '' } })}>使用</ButtonBase>
            </SkillButton>
          </Skill>
        ))}
      </div>
    </StoryNpcWrapper>
  )
}

const StoryNpcWrapper = styled.div``
const Info = styled.div`
display: flex;
align-items: center;
height: fit-content;
flex-wrap: wrap;
& > *:not(:last-child) {
  margin-right: ${space.s};
}
`
const Status = styled.div`
display: flex;
align-items: center;
`
const Icon = styled.img`
width: 1.6rem;
margin-right: ${space.xxs};
`
const HpInput = styled.input``
const ToggleButton = styled.div`
width: 10rem;
height: 1.2rem;
`
const Skill = styled.div`
display: flex;
align-items: center;
flex-wrap: wrap;
& > * {
  margin-top: ${space.s};
}
`
const SkillRow = styled.div`
display: flex;
align-items: center;
min-width: 14rem;
flex-shrink: 0;
`
const SkillName = styled.div`
margin-right: ${space.xs};
`
const SkillDepth = styled.div`
display: flex;
flex-shrink: 0;
`
const SkillDescription = styled.p`
margin: ${space.s};
margin-bottom: 0;
`
const SkillButton = styled.div`
width: 8rem;
flex-shrink: 0;
`