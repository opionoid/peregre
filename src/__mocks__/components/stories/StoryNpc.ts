import { IStoryNpcProps } from 'src/components/stories/StoryNpc';

const MOCK_STORY_NPC_PROPS: IStoryNpcProps = {
  name: '敵の名前',
  hp: 10,
  skills: [{
    name: 'なんか長そうな名前です',
    description: 'このスキルを使用したとき何かを発動する。何かを発動して敵が何かの状態であれば、何かの追加効果を発生させる。',
    shouldCast: false,
    depth: 0,
  }]
};

export default MOCK_STORY_NPC_PROPS;
