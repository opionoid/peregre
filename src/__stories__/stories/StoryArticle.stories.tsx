import { Story, Meta } from '@storybook/react'
import { StoryArticle, IStoryArticleProps } from 'src/components/stories/StoryArticle'
import MOCK_STORY_ARTICLE_PROPS from 'src/__mocks__/components/stories/StoryArticle'

export default {
  title: 'components/stories/StoryArticle',
  component: StoryArticle,
} as Meta

const Template: Story<IStoryArticleProps> = (args) => <StoryArticle {...args} />
export const Default = Template.bind({})
Default.args = {
  ...MOCK_STORY_ARTICLE_PROPS
}