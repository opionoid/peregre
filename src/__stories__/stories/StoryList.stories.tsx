import { Story, Meta } from '@storybook/react'
import { StoryList, IStoryListProps } from 'src/components/stories/StoryList'
import MOCK_STORY_LIST_PROPS from 'src/__mocks__/components/stories/StoryList'

export default {
  title: 'components/stories/StoryList',
  component: StoryList,
} as Meta

const Template: Story<IStoryListProps> = (args) => <StoryList {...args} />
export const Default = Template.bind({})
Default.args = {
  ...MOCK_STORY_LIST_PROPS,
}
