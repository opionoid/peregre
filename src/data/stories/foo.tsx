import { IStoryArticleProps, StoryArticle } from "src/components/stories/StoryArticle"

const info: Omit<IStoryArticleProps, 'children'> = {
  title: '暗狐行燈物語',
  thumbnail: { src: '', alt: '' },
  publishDate: '2021年11月21日',
  people: { from: 2, to: 3 },
  time: 4,
  difficulty: '標準',
  tags: ['和風'],
  abstract: '',
}

const content = <>
  <p>
    てすとです
  </p>
</>

export const Story = <StoryArticle {...info}>{content}</StoryArticle>