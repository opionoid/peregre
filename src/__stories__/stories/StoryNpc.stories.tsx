import { Story, Meta } from '@storybook/react'
import { RecoilRoot } from 'recoil'
import { StoryNpc, IStoryNpcProps } from 'src/components/stories/StoryNpc'
import MOCK_STORY_NPC_PROPS from 'src/__mocks__/components/stories/StoryNpc'

export default {
  title: 'components/stories/StoryNpc',
  component: StoryNpc,
} as Meta

const Template: Story<IStoryNpcProps> = (args) => <RecoilRoot><StoryNpc {...args} /></RecoilRoot>
export const Default = Template.bind({})
Default.args = {
  ...MOCK_STORY_NPC_PROPS
}