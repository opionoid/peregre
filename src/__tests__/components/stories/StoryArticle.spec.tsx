import { StoryArticle } from 'src/components/stories/StoryArticle';
import MOCK_STORY_ARTICLE_PROPS from 'src/__mocks__/components/stories/StoryArticle';
import { render } from '@testing-library/react';

describe('stories', () => {
  const snapshot = render(<StoryArticle {...MOCK_STORY_ARTICLE_PROPS} />);

  it('default test', () => {
    expect(snapshot.baseElement).toMatchSnapshot();
  });
});
