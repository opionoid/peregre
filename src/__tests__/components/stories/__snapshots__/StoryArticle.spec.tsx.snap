// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stories default test 1`] = `
<body>
  <div>
    <article
      class="sc-gsDJrp jycdiW"
    >
      <img
        alt=""
        class="sc-dkPtyc jTnJlB"
        src="https://picsum.photos/id/1037/400/300"
      />
      <h1
        class="sc-hKwCoD kXtqUB"
      >
        タイトル
      </h1>
      <p
        class="sc-eCImvq eXSYDF"
      >
        ことれの
      </p>
      <ul
        class="sc-jRQAMF dIEAXp"
      >
        <li
          class="sc-gKckTs jaqPkv"
        >
          <img
            class="sc-iCfLBT lpnpD"
            decoding="async"
            src="Person.svg"
          />
          2
          ~
          3
        </li>
        <li
          class="sc-gKckTs jaqPkv"
        >
          <img
            class="sc-iCfLBT lpnpD"
            decoding="async"
            src="Time.svg"
          />
          4
          h
        </li>
        <li
          class="sc-gKckTs jaqPkv"
        >
          <img
            alt="難易度"
            class="sc-iCfLBT lpnpD"
            decoding="async"
            src="MaxDepth.svg"
          />
          標準
        </li>
      </ul>
      <p
        class="sc-furvIG gJtkME"
      >
        テキストが入ります。テキストが入ります。テキストが入ります。テキストが入ります。テキストが入ります。テキストが入ります。テキストが入ります。
      </p>
      <hr />
      <div
        aria-hidden="false"
      >
        <button
          aria-disabled="false"
          class="sc-bdvvaa gazzRl"
          type="button"
        >
          書を読み進める
        </button>
      </div>
      <div
        aria-hidden="true"
        id="story-contents"
      >
        サンプル
      </div>
    </article>
  </div>
</body>
`;
